scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -649,294 +661,134 @@ def dijkstra(csgraph, directed=True, indices=None,
-                                                       sizeof(FibonacciNode))
-    if nodes == NULL:
-        raise MemoryError("Failed to allocate memory in "
-                          "_dijkstra_directed_multi")
-
-    # initialize the heap with each of the starting
-    # nodes on the heap and in a scanned state with 0 values
-    # and their entry of the distance matrix = 0
-    # pred will lead back to one of the starting indices
-    _dijkstra_setup_heap_multi(&heap, nodes, source_indices,
-                               sources, dist_matrix, return_pred)
-
-    while heap.min_node:
-        v = remove_min(&heap)
-        v.state = SCANNED
-
-        _dijkstra_scan_heap_multi(&heap, v, nodes,
-                                  csr_weights, csr_indices, csr_indptr,
-                                  pred, sources, return_pred, limit)
-
-        # v has now been scanned: add the distance to the results
-        dist_matrix[v.index] = v.val
-
-    free(nodes)
-    return 0
-
-@cython.boundscheck(False)
-cdef int _dijkstra_undirected(
-            const int[:] source_indices,
-            const double[:] csr_weights,
-            const int[:] csr_indices,
-            const int[:] csr_indptr,
-            const double[:] csrT_weights,
-            const int[:] csrT_indices,
-            const int[:] csrT_indptr,
-            double[:, :] dist_matrix,
-            int[:, :] pred,
-            DTYPE_t limit) except -1:
-    cdef:
-        unsigned int Nind = dist_matrix.shape[0]
-        unsigned int N = dist_matrix.shape[1]
-        unsigned int i, k, j_source
-        int return_pred = (pred.size > 0)
-        FibonacciHeap heap
-        FibonacciNode *v
-        FibonacciNode* nodes = <FibonacciNode*> malloc(N *
-                                                       sizeof(FibonacciNode))
-    if nodes == NULL:
-        raise MemoryError("Failed to allocate memory in _dijkstra_undirected")
+        # pairs of {-distance, vertex index} will be pushed
+        # to treat it as a min-heap instead of max-heap
+        dijkstra_queue_t heap = dijkstra_queue_t()
+        dist_index_pair_t v
 
     for i in range(Nind):
         j_source = source_indices[i]
+        dist_matrix[j_source] = 0
+        heap.push(dist_index_pair_t(-dist_matrix[j_source], j_source))
+        if return_sources:
+            sources[j_source] = j_source
 
-        for k in range(N):
-            initialize_node(&nodes[k], k)
-
-        dist_matrix[i, j_source] = 0
-        heap.min_node = NULL
-        insert_node(&heap, &nodes[j_source])
-
-        while heap.min_node:
-            v = remove_min(&heap)
-            v.state = SCANNED
-
-            _dijkstra_scan_heap(&heap, v, nodes,
-                                csr_weights, csr_indices, csr_indptr,
-                                pred, return_pred, limit, i)
+    if return_pred:
+        assert pred.size == N
+    if return_sources:
+        assert sources.size == N
```

Comment:
We don't use ``assert`` in production code. This is because the Python interpreter will ignore them if ``__debug__`` is not True.

Again, ``.size`` requires interaction with the Python interpreter. ``.shape[0]`` does not.